nrf_wifi/fw_if/umac_if/src/event.c

```diff
@@ -927,62 +927,94 @@ static enum nrf_wifi_status umac_process_sys_events(struct nrf_wifi_fmac_dev_ctx
-
 			mode_event = (struct nrf_wifi_event_raw_config_mode *)sys_head;
+			struct tx_config *config = &def_dev_ctx->tx_config;
+			unsigned char if_idx = mode_event->if_index;
+			struct nrf_wifi_fmac_vif_ctx *vif = def_dev_ctx->vif_ctx[if_idx];
+
 			if (!mode_event->status) {
-				def_dev_ctx->vif_ctx[mode_event->if_index]->mode =
-								mode_event->op_mode;
-				/**
-				 * Set the transmit queue for RAW packet transmission
-				 */
+				vif->mode = mode_event->op_mode;
 #ifdef CONFIG_NRF700X_RAW_DATA_TX
-				if (mode_event->op_mode == (NRF_WIFI_TX_INJECTION_MODE |
-							    NRF_WIFI_STA_MODE)) {
-					def_dev_ctx->tx_config.peers[MAX_PEERS].peer_id =
-										MAX_PEERS;
-					def_dev_ctx->tx_config.peers[MAX_PEERS].if_idx =
-									mode_event->if_index;
-					def_dev_ctx->vif_ctx[mode_event->if_index]->if_type =
-									NRF_WIFI_STA_TX_INJECTOR;
-					def_dev_ctx->vif_ctx[mode_event->if_index]->txinjection_mode
-									= true;
-				} else
+				vif->txinjection_mode = false;
 #endif /* CONFIG_NRF700X_RAW_DATA_TX */
-				if (mode_event->op_mode == NRF_WIFI_STA_MODE) {
-					def_dev_ctx->vif_ctx[mode_event->if_index]->if_type =
-									NRF_WIFI_IFTYPE_STATION;
+#ifdef CONFIG_NRF700X_PROMISC_DATA_RX
+				vif->promisc_mode = false;
+#endif /* CONFIG_NRF700X_PROMISC_DATA_RX */
+				if ((mode_event->op_mode & NRF_WIFI_STA_MODE)
+								== NRF_WIFI_STA_MODE) {
+					mode_event->op_mode ^= NRF_WIFI_STA_MODE;
+					vif->if_type = NRF_WIFI_STA_MODE;
+					config->peers[MAX_PEERS].peer_id = -1;
+					config->peers[MAX_PEERS].if_idx = -1;
+
+#if defined(CONFIG_NRF700X_RAW_DATA_TX) && defined (CONFIG_NRF700X_PROMISC_DATA_RX)
+					if (mode_event->op_mode) {
+						if ((mode_event->op_mode ^
+						    (NRF_WIFI_PROMISCUOUS_MODE |
+						     NRF_WIFI_TX_INJECTION_MODE)) == 0) {
+							vif->if_type =
+								NRF_WIFI_STA_PROMISC_TX_INJECTOR;
+							config->peers[MAX_PEERS].peer_id
+									= MAX_PEERS;
+							config->peers[MAX_PEERS].if_idx
+									= if_idx;
+							vif->txinjection_mode = true;
+							vif->promisc_mode = true;
+							break;
+						}
+#endif /* CONFIG_NRF700X_RAW_DATA_TX && defined CONFIG_NRF700X_PROMISC_DATA_RX */
 #ifdef CONFIG_NRF700X_RAW_DATA_TX
-					def_dev_ctx->vif_ctx[mode_event->if_index]->txinjection_mode
-									= false;
+						if ((mode_event->op_mode ^
+						     NRF_WIFI_TX_INJECTION_MODE) == 0) {
+							config->peers[MAX_PEERS].peer_id
+									= MAX_PEERS;
+							config->peers[MAX_PEERS].if_idx
+									= if_idx;
+							vif->if_type = NRF_WIFI_STA_TX_INJECTOR;
+							vif->txinjection_mode = true;
+							break;
+						}
 #endif /* CONFIG_NRF700X_RAW_DATA_TX */
-					def_dev_ctx->tx_config.peers[MAX_PEERS].peer_id = -1;
+#ifdef CONFIG_NRF700X_PROMISC_DATA_RX
+						if ((mode_event->op_mode ^
+						     NRF_WIFI_PROMISCUOUS_MODE) == 0) {
+							vif->if_type = NRF_WIFI_STA_PROMISC;
+							vif->promisc_mode = true;
+							break;
+						}
+					}
+#endif /* CONFIG_NRF700X_PROMISC_DATA_RX */
```

Comment:
This requires some code refactoring, and would like to be moved to a dedicated routine even before another indentation in this commit.

Reply:
I have moved the functionality to a dedicated routine. The Code has multiple CONFIG checks which are also refactored as much as possible to make sure that the functionality is not affected.